### PR TITLE
Add AdamW optimizer to extension/training

### LINKED ISSUE #18766: Support for AdamW optimizer in ExecuTorch
As of now, ExecuTorch supports SGD as the only optimizer, with Adam listed as planned.

Given that AdamW is the default optimizer used in most modern PyTorch training workflows, is AdamW support also expected as part of the ExecuTorch optimizer roadmap?

Are there any known design or semantic considerations in existing AdamW implementations (e.g., weight decay behavior or optimizer state handling) that influence how AdamW support is approached in ExecuTorch?

Thanks!

cc @JacobSzwejbka

### PR DESCRIPTION
**Pull request by BryanBradfo**
Adds AdamW to the training optimizer extension. It's a port of the existing SGD implementation at `extension/training/optimizer/sgd.{h,cpp}`, with the main algorithmic difference being decoupled weight decay (the parameter gets decayed directly instead of mixing the decay into the gradient). Matches `torch.optim.AdamW` with default settings.

Fixes #18766

### Scope

C++ only for this PR. Python bindings are left out on purpose: the pybindings file has a TODO to build a generic optimizer interface first, so copying `PySGD` to `PyAdamW` now would just add duplication. Happy to follow up with that. `amsgrad` and `maximize` are also left out, both rarely used and easy to add later if needed.

### Test plan

Six new gtests pass, and the SGD regression stays green:

    $ buck2 test //extension/training/optimizer/test:adamw_test
    [  PASSED  ] 6 tests.

    $ buck2 test //extension/training/optimizer/test:sgd_test
    [  PASSED  ] 5 tests.

Output was also cross-checked against `torch.optim.AdamW` on four small cases (simple convergence, decoupled weight decay, multi-parameter). All four match to six decimal places.

cc @JacobSzwejbka

### CODE REVIEW

**Comment by JacobSzwejbka** [comment]
You can ignore copilot probably its noise ratio is pretty bad. Ive been trying to figure out how to turn it off for the repo.

File: extension/training/optimizer/adamw.h [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+   * fully qualified name.
+   */
+  ::executorch::runtime::Error step(
+      const std::map<std::string_view, executorch::aten::Tensor>&
+          named_gradients);
+
+ private:
+  std::vector<AdamWParamGroup> param_groups_;
+  std::unordered_map<void*, std::unique_ptr<AdamWParamState>> state_;
+  std::unique_ptr<AdamWOptions> defaults_;
```
Comment (JacobSzwejbka) [inline extension/training/optimizer/adamw.h]: Why wrap them in a ptr?

File: extension/training/optimizer/adamw.cpp [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        std::memset(v_buf_ptr, 0, g.nbytes());
+
+        Tensor m_buf(nullptr);
+        Tensor v_buf(nullptr);
+#ifdef USE_ATEN_LIB
+        std::vector<int64_t> sizes(g.sizes().begin(), g.sizes().end());
+        m_buf = torch::from_blob(m_buf_ptr, sizes, g.scalar_type());
+        v_buf = torch::from_blob(v_buf_ptr, sizes, g.scalar_type());
+#else
+        TensorImpl* m_impl = new TensorImpl(
```
Comment (JacobSzwejbka) [inline extension/training/optimizer/adamw.cpp]: you can use extension/tensor to get rid of a lot of boilerplate cruft here

File: extension/training/optimizer/adamw.cpp [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+using ::executorch::runtime::Error;
+
+namespace executorch {
+namespace extension {
+namespace training {
+namespace optimizer {
+
+namespace {
+// out[i] = a[i] + alpha * b[i]
+void add_out_hack(
```
Comment (JacobSzwejbka) [inline extension/training/optimizer/adamw.cpp:29]: ahh my legacy... @manuelcandales  do you remember how to call executorch ops outside of the interpreter. I cant recall

**Review by JacobSzwejbka — CHANGES_REQUESTED** [review]

File: extension/training/optimizer/adamw.cpp [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+            g.sizes().size(),
+            const_cast<TensorImpl::SizesType*>(g.sizes().data()),
+            m_buf_ptr,
+            const_cast<TensorImpl::DimOrderType*>(g.dim_order().data()));
+        TensorImpl* v_impl = new TensorImpl(
+            g.scalar_type(),
+            g.sizes().size(),
+            const_cast<TensorImpl::SizesType*>(g.sizes().data()),
+            v_buf_ptr,
+            const_cast<TensorImpl::DimOrderType*>(g.dim_order().data()));
```
[the remark this answers is not in the document]
Comment (BryanBradfo) [reply extension/training/optimizer/adamw.cpp]: Good catch in principle. This matches the existing pattern in SGD (sgd.cpp:119-125) which constructs TensorImpl with the same d_p.sizes().data() / d_p.dim_order().data() pointers from the gradient. In practice, parameter shapes are stable across training steps in ExecuTorch's captured graph model. Happy to address both optimizers together in a follow-up if the maintainer prefers.

File: extension/training/optimizer/adamw.cpp [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+      mul_out_hack(exp_avg, beta1, exp_avg);
+      add_out_hack(exp_avg, g, 1.0 - beta1, exp_avg);
+
+      // Second moment: v <- beta2 * v + (1 - beta2) * g^2
+      addcmul_sq_out_hack(exp_avg_sq, g, beta2, exp_avg_sq);
+
+      // Bias-corrected update.
+      const double bias_correction1 = 1.0 - std::pow(beta1, step);
+      const double bias_correction2 = 1.0 - std::pow(beta2, step);
+      adamw_update_hack(
```
[the remark this answers is not in the document]
Comment (BryanBradfo) [reply extension/training/optimizer/adamw.cpp:205]: SGD also has no parameter validation (sgd.h), and PyTorch's C++ AdamW documents 0 <= beta < 1 as a precondition rather than a runtime check. Adding validation only to AdamW would be inconsistent. Happy to add range checks across both optimizers in a follow-up if the maintainer wants.

File: extension/training/optimizer/adamw.h (hunk 1 above)
Comment (BryanBradfo) [reply extension/training/optimizer/adamw.h]: I copied it from SGD's pattern. You're right, defaults_ is never null so it doesn't need a pointer. I can change it to store by value

File: extension/training/optimizer/adamw.cpp (hunk 2 above)
Comment (BryanBradfo) [reply extension/training/optimizer/adamw.cpp]: yes, i'll switch the state allocation to use it, which also simplifies the destructor. Thanks for the pointer!

**Review by JacobSzwejbka — APPROVED** [review]

**Comment by psiddh** [comment]
@BryanBradfo Can you complete the EasyCLA CI step please ? Once completed, we can merge this PR.
Thank you again for your contribution

**Comment by psiddh** [comment]
@BryanBradfo Pinging again. Could you pls take care of easy CLA ?

**Comment by psiddh** [comment]
Few compilation issues..

 /__w/executorch/executorch/pytorch/executorch/extension/training/optimizer/adamw.cpp:168:22: error: no matching function for call to 'make_tensor_ptr'
              auto m_ptr = make_tensor_ptr(
                           ^~~~~~~~~~~~~~~
      /__w/executorch/executorch/pytorch/executorch/../executorch/extension/tensor/tensor_ptr.h:81:18: note: candidate function not viable: no known conversion from 'executorch::runtime::TensorShapeDynamism' to 'executorch::aten::Device' (aka 'executorch::runtime::etensor::Device') for 4th argument
      inline TensorPtr make_tensor_ptr(
                       ^

**Comment by psiddh** [comment]
Pushed two commits to this branch (`maintainer_can_modify` was on, so no new PR needed): a merge of current `main`, and a fix for the build break. Your commits are untouched.

**The compile error.** `make_tensor_ptr` gained a `Device` parameter after this branch was written, so the two state-buffer calls were passing `TensorShapeDynamism::STATIC` where a `Device` is expected and the deleter where the dynamism is expected. Both moment buffers come from `malloc`, so they get an explicit CPU device. I also added the null check Copilot flagged on those same two allocations — they were used unconditionally, so a failure dereferenced null instead of returning an error. The other Copilot points (dtype/shape validation, parameter range checks) I left alone; your SGD-parity argument holds and @JacobSzwejbka said to discount Copilot.

**Verification.** I reproduced the exact CI error locally, confirmed the fix clears it, and `lintrunner` is clean. I could not run `adamw_test` here — this machine SIGKILLs freshly built binaries — so instead I transcribed `AdamW::step`'s arithmetic verbatim (including the float round-trips) and diffed it against `torch.optim.AdamW` over six configurations:

```
case                                adamw.cpp        torch     absdiff
Simple  (wd=0, g=-1, 10 steps)      2.0000002    2.0000002    0.00e+00
DecoupledWD (g=0, wd=0.5, 1)        0.9500000    0.9500000    0.00e+00
MultiParam p1 (g=-1, 5 steps)       1.5000001    1.5000001    0.00e+00
MultiParam p2 (g=+1, 5 steps)       1.4999999    1.4999999    0.00e+00
Both wd and grad nonzero            2.9613018    2.9613023    4.77e-07
Nondefault betas                   -0.2605234   -0.2605233    2.98e-08
```

The decoupled-weight-decay ordering matches, and all four gtest expectations agree with PyTorch. Please still run `buck2 test //extension/training/optimizer/test:adamw_test` on your side to confirm the real binary.

**Still blocking: EasyCLA.** Commits `5f1c90c` and `bda405f` are not authorized under a signed CLA, so this cannot merge regardless of CI. @BryanBradfo that one needs you — https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/29311126/463628495/18848/#/?version=2

cc @JacobSzwejbka

The fix commit was authored with Claude Code.

**Comment by psiddh** [comment]
Opened #22857 carrying this work forward, with credit to @BryanBradfo in both the commit message and the PR description.

To be explicit about why it is a squash under a different author rather than a re-post of your commits: EasyCLA keys off commit authorship, so any PR containing `5f1c90c` and `bda405f` fails the same check. That is the only reason the authorship differs — the implementation, the tests and the design calls are all yours.

@BryanBradfo if you sign the CLA, this PR is the better one to land and I will close #22857. The build fix you need is already pushed here, and CI is green apart from the CLA gate:
https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/29311126/463628495/18848/#/?version=2

cc @JacobSzwejbka — your call which one you would rather take.

**Comment by psiddh** [comment]
## Upgraded verification: the real C++ now runs, not just a transcription

Earlier I could only say I had transcribed `AdamW::step()`'s arithmetic and diffed it against
`torch.optim.AdamW`, because the gtest binary was being SIGKILLed in my environment. I got the
actual compiled ExecuTorch code running, so this is now direct evidence.

```
case                                    got     expected     |diff|  result
Simple (wd=0,g=-1,x10)           2.00000024   2.00000000   2.38e-07  PASS
DecoupledWD (g=0,wd=.5,x1)       0.94999999   0.94999999   0.00e+00  PASS
MultiParam p1 (g=-1,x5)          1.50000012   1.50000000   1.19e-07  PASS
MultiParam p2 (g=+1,x5)          1.49999988   1.50000000   1.19e-07  PASS
Varying wd=0 (8 steps)           0.84547687   0.84547687   0.00e+00  PASS
Varying wd=0.1 (8 steps)         0.77574724   0.77574724   0.00e+00  PASS
6/6 passed
```

The two varying-gradient values are the ones taken from `torch.optim.AdamW`, and the real
implementation reproduces them exactly.

Per-parameter state isolation, which the cases above do not cover because they use one optimizer
each, checked separately with two parameters in a single optimizer:

```
two params, one optimizer, 5 steps:
  a: 1.50000012 (expect ~1.5) PASS
  b: 1.49999988 (expect ~1.5) PASS
  partial gradients: b unchanged = PASS (1.49999988 -> 1.49999988)
```

That last line also covers the `named_gradients.find(...) == end()` skip path, and the whole run
exercises the state-allocation reorder from the third commit, since state is allocated on step 1
and reused on steps 2-5.

### What this still does not cover

The gtest harness itself (`TensorFactory`, the `ET_EXPERIMENTAL` wrappers) and the Buck target. My
driver calls `AdamW::step()` directly against `TensorImpl`. The gtest binary is still SIGKILLed
here by what looks like an endpoint-security rule — a trivial binary linking the same objects runs
fine, so it is not the code. Someone should still run
`buck2 test //extension/training/optimizer/test:adamw_test` before merge.
